packages/thumbprint-react/components/Modal/components/transition.tsx

```diff
@@ -1,37 +1,39 @@
 import React from 'react';
-import PropTypes from 'prop-types';
 
 const EXITED = 'exited';
 const ENTERING = 'entering';
 const ENTERED = 'entered';
 const EXITING = 'exiting';
 
-const propTypes = {
-    children: PropTypes.func.isRequired,
-    in: PropTypes.bool.isRequired,
-    timeout: PropTypes.shape({
-        enter: PropTypes.number.isRequired,
-        exit: PropTypes.number.isRequired,
-    }).isRequired,
-    onEntered: PropTypes.func,
-    onExited: PropTypes.func,
-};
-
-const defaultProps = {
-    onEntered: undefined,
-    onExited: undefined,
-};
-
-class Transition extends React.Component {
-    constructor(props) {
+type Stage = 'exited' | 'entering' | 'entered' | 'exiting' | null;
+
+interface PropTypes {
+    children: (stage: Stage) => JSX.Element;
+    in: boolean;
+    timeout: {
+        enter: number;
+        exit: number;
+    };
+    onEntered?: () => void;
+    onExited?: () => void;
+}
+
+interface StateTypes {
+    stage: Stage;
+}
+
+export default class Transition extends React.Component<PropTypes, StateTypes> {
+    currentTimeout: number | null = null;
```

Comment:
Could this go in the `constructor`? I'd rather avoid adding a Babel plugin in this PR if we don't need.

Reply:
I fixed this in the end by storing these things on `this.state` instead of `this`, which I think was always what we should have been doing.

If you want to put things on `this` directly, TS requires that you declare them as class properties first, but it looks like the TS-Babel plugin doesn't support that syntax directly. We haven't run across this so far because I already did most of the work to convert all the other components to function components.

A followup task is probably to convert the Modal to a function component too, but for now storing these values on `this.state` allows us to complete this TS migration without adding a new Babel plugin.

Reply:
Ah, interesting.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes

Based on that MDN link, would `Transition.currentTimeout = 123` outside of the classs also be a way of declaring a class property? (No need to change the code, just curious.)

Reply:
no, you would also need to extend the class' type definition before you could add that property to it.